pkg/client/event_deduplicator.go

```diff
@@ -0,0 +1,188 @@
+package client
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	chain "github.com/keep-network/keep-ecdsa/pkg/chain"
+)
+
+// eventDeduplicator decides whether the given event should be handled by the
+// client or not.
+//
+// Event subscription may emit the same event two or more times. The same event
+// can be emitted right after it's been emitted for the first time. The same
+// event can also be emitted a long time after it's been emitted for the first
+// time. It is eventDeduplicator's responsibility to decide whether the given
+// event is a duplicate and should be ignored or if it is not a duplicate and
+// should be handled.
+//
+// Four events are supported:
+// - key generation request for a new keep,
+// - signature request for a keep,
+// - keep close request,
+// - keep terminate request.
+type eventDeduplicator struct {
```

Comment:
Wouldn't `eventApprover` (or something similar) reflect better what are we doing here? The real deduplication occurs in `keepEventTrack` instances. Here we are one level upper as we're also checking things against the chain.

Reply:
I'd say that real deduplication happens in `Deduplicator` (former `eventDeduplicator`). `keepEventTrack` is filtering out *immediate* duplicates but duplicates that will be pulled by subscription's past events call, will be deduplicated in `Deduplicator`.

Reply:
Ok, that makes sense now. Thanks for the clarification.